### PR TITLE
Deep and shallow copy

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Add shallow and deep copy

## How was this patch tested?
unit test & model validation

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/1173

### CODE REVIEW

**Comment by wzhongyuan** [comment]
PR validation passed http://172.168.2.101:8080/job/BigDL-PR-V/697/

**Comment by wzhongyuan** [comment]
@yangw1234  @i8run   please help review the changes, thanks

**Comment by wzhongyuan** [comment]
Just ran another latest PR job sucessfully
http://172.168.2.101:8080/job/BigDL-PR-V/801/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -448,10 +449,93 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+      .asInstanceOf[AbstractModule[A, B, T]]
+    setWeightAndBias(copy, deepCopy)
+    copy
+  }
+
+
+  private def resetStorage(storages: mutable.HashMap[Int, Any]):
+    mutable.HashMap[Int, Any] = {
+    val storages = new mutable.HashMap[Int, Any]()
+    storages
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Two questions. 1. Why `reset` will return a new map here? 2. Have we used this method `resetStorage`?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 2]
```diff
@@ -448,10 +449,93 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
     SerializationUtils.clone(this)
   }
 
+  def clone(deepCopy : Boolean = true): AbstractModule[A, B, T] = {
+    val moduleData = ModuleData[T](this.
+      asInstanceOf[AbstractModule[Activity, Activity, T]], Seq[String](), Seq[String]())
+    val storages = new mutable.HashMap[Int, Any]()
+    val context = SerializeContext(moduleData, storages, ProtoStorageType, false)
+    val serializedModule = ModuleSerializer.serialize[T](context)
+      .bigDLModule
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Style. To single line.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 3]
```diff
@@ -448,10 +449,93 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+      storages, ProtoStorageType, false)
+    ModuleLoader.initTensorStorage[T](deserializeContext)
+    val copy = ModuleSerializer.load[T](deserializeContext).module
+      .asInstanceOf[AbstractModule[A, B, T]]
+    setWeightAndBias(copy, deepCopy)
+    copy
+  }
+
+
+  private def resetStorage(storages: mutable.HashMap[Int, Any]):
```
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: It seems this method is not used anywhere

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 4]
```diff
@@ -448,10 +449,93 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+          setLayerWeightAndBias(params,
+            copiedModuleParamTable.get(name).get.asInstanceOf[Table], deepCopy)
+      }
+    }
+  }
+
+  private def setLayerWeightAndBias(params : Table,
+                                    copyParams : Table, deepCopy : Boolean): Unit = {
+
+    copyParam(params, copyParams, deepCopy, "weight")
```
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Maybe there are parameters other than "weight" and "bias" like the "runningMean" and "runningVar" in BatchNormalization

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/AbstractModuleSpec.scala [hunk 5]
```diff
@@ -327,4 +329,97 @@ class AbstractModuleSpec extends FlatSpec with Matchers {
...
+    val shallowCopy = linear.clone(false).asInstanceOf[Linear[Float]]
+
+    val originWeight = linear.weight
+
+    val originBias = linear.bias
+
+    originWeight.fill(1.0f)
+    originBias.fill(2.0f)
+
+    val input = Tensor[Float](2, 2)
```
Comment (i8run) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/AbstractModuleSpec.scala]: The `input` should be initialized with random value.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/AbstractModuleSpec.scala [hunk 6]
```diff
@@ -327,4 +329,97 @@ class AbstractModuleSpec extends FlatSpec with Matchers {
...
+
+  }
+
+  "Deep copy" should  "work properly" in {
+
+    val linear = Linear[Float](2, 2)
+
+    val deepCopy = linear.clone(true).asInstanceOf[Linear[Float]]
+
+    val input = Tensor[Float](2, 2)
```
Comment (i8run) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/AbstractModuleSpec.scala]: The `input` should be initialized with random value.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/AbstractModuleSpec.scala [hunk 7]
```diff
@@ -327,4 +329,97 @@ class AbstractModuleSpec extends FlatSpec with Matchers {
...
+      2.0f, 3f,
+      4f, 5f
+    )
+
+    val biasData = Array(0.0f, 0.1f)
+
+    val input = Tensor[Float](2, 2).apply1(_ => Random.nextFloat())
+    val weight = Tensor[Float](Storage(kernelData), 1, Array(outputSize, inputSize))
+    val bias = Tensor[Float](Storage(biasData), 1, Array(outputSize))
+    val linear = quantized.Linear[Float](outputSize, inputSize, initWeight = weight,
```
Comment (i8run) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/AbstractModuleSpec.scala]: The `input`, `weight`, `bias` should be initialized with random value.

**Comment by wzhongyuan** [comment]
@i8run  @yangw1234  PR validation passed http://172.168.2.101:8080/job/BigDL-PR-V/837/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 8]
```diff
@@ -458,6 +459,81 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
     SerializationUtils.clone(this)
   }
 
+  def clone(deepCopy : Boolean): AbstractModule[A, B, T] = {
```
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: name of this method is conflict with java Object.clone()

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: I removed the default the value ,  it should be ok now , previously `def clone(deepCopy : Boolean = true):`

Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Sorry, I did not notice. It should work.

**Comment by wzhongyuan** [comment]
@i8run  @yangw1234   Ran another PR successfully http://172.168.2.101:8080/job/BigDL-PR-V/848/  please help  check and let me know if you have other concerns

**Pull request merged**